File: javascript/rover.js

// VARIABLE DECLARATION
const DIRECTION = {
  N: "N",
  E: "E",
  S: "S",
  W: "W"
}

const MOVEMENTS = {
  f: "f",
  b: "b",
  l: "l",
  r: "r"
}

const f = MOVEMENTS.f;
const b = MOVEMENTS.b;
const l = MOVEMENTS.l;
const r = MOVEMENTS.r;

const N = DIRECTION.N;
const E = DIRECTION.E;
const S = DIRECTION.S;
const W = DIRECTION.W;

const Rover = (facing, x, y) => {

  const turnLeftFacing = () => {
    switch (facing) {
      case N:
        facing = W;
        break;
      case E:
        facing = N;
        break;
      case S:
        facing = E;
        break;
      case W:
        facing = S;
        break;
    }
  }
  const turnRightFacing = () => {
    switch (facing) {
      case N:
        facing = E;
        break;
      case E:
        facing = S;
        break;
      case S:
        facing = W;
        break;
      case W:
        facing = N;
        break;
    }
  }

  const turn = (command) => (command === l ? turnLeftFacing : turnRightFacing)();

  const move = (command) => {
    const incrementer = command === b ? -1 : 1;
    switch (facing) {
      case N:
        y -= incrementer;
        break;
      case S:
        y += incrementer;
        break;
      case E:
        x += incrementer;
        break;
      case W:
        x -= incrementer;
        break;
      default:
        throw new Error("wrong facing -.-'");
    }
  };

  return {
    execute(commands) {
      commands.forEach((command) => {
        switch (command) {
          case l:
          case r:
            turn(command);
            break;
          case f:
          case b:
            move(command);
            break;
          default:
            throw new Error("wrong command -.-'");
        }
      });
      return { facing, x, y };
    },

    get x() {
      return x;
    },
    get y() {
      return y;
    },
    get facing() {
      return facing;
    },
  };
};

module.exports = {Rover, f, b, l, r, N, E, S, W};

File: java_environment/src/test/java/rover/Rover.java
package rover;

import java.util.Arrays;

public class Rover {
    private String[] directionArray = { "N", "E", "S", "W" };

    public Rover(String Facing, int X, int Y) {
        facing = Facing;
        x = X;
        y = Y;
    }

    private String facing;

    public String getFacing() {
        return facing;
    }

    private int x;
    public int getX() {
        return x;
    }
    private int y;
    public int getY() {
        return y;
    }

    public Rover execute(String[] commands) {
        for (String command : commands) {
            switch (command) {
                case "l":
                    facing = directionArray[getPreviousDirectionIndex(facing)];
                    break;
                case "r":
                    facing = directionArray[getNextDirectionIndex(facing)];
                    break;
                    case "f":
                    case "b":
                      move(command);
                      break;
                default:
                    break;
            }
        }
        return this;
    }

    private void move(String command) {
        int incrementer = command == "b" ? -1 : 1;
        switch (facing) {
          case "N":
            y -= incrementer;
            break;
          case "S":
            y += incrementer;
            break;
          case "E":
            x += incrementer;
            break;
          case "W":
            x -= incrementer;
            break;
          default:
            throw new Error("wrong facing -.-'");
        }
    }

    private int indexOf(String c) {
        return Arrays.asList(directionArray).indexOf(c);
    }

    private int getIndexOfMinusOne(String direction) {
        int returner = indexOf(direction) - 1;
        return returner;
    }

    private int getIndexOfPlusOne(String direction) {
        int returner = indexOf(direction) + 1;
        return returner;
    }

    private int normalizeDirectionIndex(int index) {
        int returner = (index + directionArray.length) % directionArray.length;
        return returner;
    }

    private int getPreviousDirectionIndex(String direction) {
        int returner = normalizeDirectionIndex(getIndexOfMinusOne(direction));
        return returner;
    }

    private int getNextDirectionIndex(String direction) {
        int returner = normalizeDirectionIndex(getIndexOfPlusOne(direction));
        return returner;
    }
}

// private class RoverReturnValues {
//     String facing;
//     int x;
//     int y;
// }